File: swift-wasm-webgl/Sources/math-and-graphics-types/Plane.swift
import JavaScriptKit

struct Plane<T: TypedArrayElement & StaticSized & Mathable & Sqrt> {
    let normal: Vector3<T>
    let d: T

    init(normal: Vector3<T>, point: Vector3<T>) {
        self.normal = normal.normalized
        d = Vector3.dot(normal, point)
    }

    /**
	Finds t such that (point + normal*t) is a point on the plane.
	*/
    func signedDistanceTo(point: Vector3<T>) -> T {
        d - Vector3.dot(normal, point)
    }

    func closestPointTo(point: Vector3<T>) -> Vector3<T> {
        normal * signedDistanceTo(point: point) + point
    }
}

File: rust-webgl/src/extra_math/plane.rs
use nalgebra::{Point3, Vector3};

use super::primitives::Float;

pub struct Plane {
    normal: Vector3<Float>,
    d: Float,
}

impl Plane {
    pub fn new_from_normal_and_point(normal: Vector3<Float>, point: &Vector3<Float>) -> Self {
        Self {
            normal,
            d: normal.dot(point),
        }
    }

    /**
    Finds t such that (point + normal*t) is a point on the plane.
    */
    pub fn signed_distance_to_point(&self, point: &Point3<Float>) -> Float {
        self.d - self.normal.dot(&point.coords)
    }

    pub fn closest_point_to_point(&self, point: &Point3<Float>) -> Vector3<Float> {
        self.normal * self.signed_distance_to_point(point) + point.coords
    }
}
